### PR TITLE
Fix pytest warnings

### LINKED ISSUE #181: Fix Unit Test Warnings
In our existing code, the unit tests yield the warnings below.  We should fix them, they don't look hard, some of them I have fixed with a line or two in course of doing other things, but I didn't merge in since it was unrelated.

Also there is an entirely new batch I've seen using slightly newer code (if you update numpy and scipy... deprecation warnings that will break code next release.... ). Our collaborators were constantly mentioning them to me, and asking if that is okay etc.

I'd rather fix the code than answer that question again.  Once we fix them all I can setup automation to error on future ones, then we stay ahead of most of it.

```
============================================================================================================== warnings summary ==============================================================================================================
tests/test_FBbasis2D.py::FBBasis2DTestCase::testFBBasis2DExpand
tests/test_FBbasis2D.py::FBBasis2DTestCase::testFBBasis2DExpand_t
tests/test_FBbasis3D.py::FBBasis3DTestCase::testFBBasis3DExpand
tests/test_FBbasis3D.py::FBBasis3DTestCase::testFBBasis3DExpand_t
tests/test_FFBbasis2D.py::FFBBasis2DTestCase::testFFBBasis2DExpand
tests/test_FFBbasis3D.py::FFBBasis3DTestCase::testFFBBasis3DExpand
tests/test_mean_estimator.py::MeanEstimatorTestCase::testEstimate
tests/test_mean_estimator.py::MeanEstimatorTestCase::testOptimize1
tests/test_mean_estimator.py::MeanEstimatorTestCase::testOptimize2
  /home/garrett/anaconda3/envs/aspire/lib/python3.6/site-packages/scipy/sparse/linalg/isolve/iterative.py:2: DeprecationWarning: scipy.sparse.linalg.cg called without specifying `atol`. The default value will be changed in a future release. For compatibility, specify a value for `atol` explicitly, e.g., ``cg(..., atol=0)``, or to retain the old behavior ``cg(..., atol='legacy')``

tests/test_FBbasis3D.py: 112 tests with warnings
tests/test_FFBbasis3D.py: 96 tests with warnings
tests/test_covar3d.py: 16 tests with warnings
tests/test_mean_estimator.py: 64 tests with warnings
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/basis/basis_utils.py:79: DeprecationWarning: np.asscalar(a) is deprecated since NumPy v1.16, use a.item() instead
    return np.asscalar(j) if scalar else j

tests/test_anisotropic_noise.py::SimTestCase::testAnisotropicNoiseVariance
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/estimation/noise.py:95: DeprecationWarning: np.asscalar(a) is deprecated since NumPy v1.16, use a.item() instead
    return np.asscalar(self.filter.evaluate(np.zeros((2, 1))))

tests/test_covar3d.py::Covar3DTestCase::testCovar3D1
tests/test_covar3d.py::Covar3DTestCase::testCovar3D2
tests/test_fourier_kernel.py::FourierKernelTestCase::test1
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/estimation/kernel.py:169: ComplexWarning: Casting complex values to real discards the imaginary part
    A[:, i] = vol_to_vec(self.convolve_volume(vec_to_vol(A[:, i])))

tests/test_mean_estimator.py::MeanEstimatorTestCase::testEstimate
tests/test_mean_estimator.py::MeanEstimatorTestCase::testOptimize1
tests/test_mean_estimator.py::MeanEstimatorTestCase::testOptimize2
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/basis/fb_3d.py:216: ComplexWarning: Casting complex values to real discards the imaginary part
    v[idx] = ang_radial.T @ x[mask]

tests/test_simulation.py::SimTestCase::testSimulationEvalCoords
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/source/simulation.py:308: DeprecationWarning: np.asscalar(a) is deprecated since NumPy v1.16, use a.item() instead
    mean_eigs_inners = np.asscalar(vol_to_vec(mean_vol).T @ vol_to_vec(eig_vols))

tests/test_simulation.py::SimTestCase::testSimulationVolCoords
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/utils/matrix.py:304: RuntimeWarning: invalid value encountered in sqrt
    return np.sqrt(ainner(x, x, axes))

tests/test_white_noise.py::SimTestCase::testWhiteNoise
  /scratch/Dropbox/PU/aspire/ASPIRE-Python.dev/src/aspire/estimation/noise.py:39: DeprecationWarning: np.asscalar(a) is deprecated since NumPy v1.16, use a.item() instead
    return np.asscalar(self.filter.evaluate(np.zeros((2, 1))))

-- Docs: https://docs.pytest.org/en/latest/warnings.html
```

### PR DESCRIPTION
**Pull request by garrettwrong**
Closes #181 .

This cleans up all but one warning which I discuss below from the associated issue. 

I didn't really feel good about the np.real() cast without looking deeper, but I can say matter-of-factually that is the operation that was happening.  I figure one of you guys can easily confirm/deny whether that was your intent before we merge it in.


The remaining warning is some code to compute the l2 norms along an axis afaict (which i think exists in np..).  Interestingly our code somehow gets negative values from a method that should be computing <x,x>, and then tries to sqrt it. I forced real, just to sanity check and get same thing, so, umm, not sure what is going on there, but I do not think it is correct. I want to confirm what is happening here, then might offer to replace with numpy.linalg.norm or something...


@janden thoughts on this snippet?

```
def anorm(x, axes=None):
    """                                                                                                                                                                                                                                       
    Calculate array norm along given axes                                                                                                                                                                                                     
    :param x: An array of arbitrary size and shape.                                                                                                                                                                                           
    :param axes: The axis along which to compute the norm. If None, the norm is calculated along all axes.                                                                                                                                    
    :return: The Euclidean (l^2) norm of x along specified axes.                                                                                                                                                                              
    """
    if axes is None:
        axes = range(x.ndim)

    # see if we have any potential for denorms etc                                                                                                                                                                                            
    smallest = np.min(np.abs(x))
    print('qqq', np.min(np.abs(x)), x.dtype)
    ## NOPE                                                                                                                                                                                                                                   

    if np.iscomplexobj(x):
        raise RuntimeError('interesting')
    print('xxx', ainner(x, x, axes))

    # wut, i need to see this again                                                                                                                                                                                                           
    x = np.real(x)
    print('yyy', ainner(x, x, axes))
    # interesting, not what I would expect from an inner product of reals. What am I missing here?                                                                                                                                            

# 2020/06/11 08:45:06 Appending a NoiseAdder to generation pipeline                                                                                                                                                                           
# qqq 3.0553412790682553e-16 float64                                                                                                                                                                                                          
# xxx [[ 6.95639929e-14 -6.85178057e-14]                                                                                                                                                                                                      
#  [-6.85178057e-14  7.19106708e-13]]                                                                                                                                                                                                         
# yyy [[ 6.95639929e-14 -6.85178057e-14]                                                                                                                                                                                                      
#  [-6.85178057e-14  7.19106708e-13]]                                                                                                                                                                                                         
# PASSED 

    return np.sqrt(ainner(x, x, axes))

```

### CODE REVIEW

**Comment by garrettwrong** [comment]
I think because `axes` is (potentially) a tuple the `ainner` is computing the inner product along two different axes of x, which can yield the negative numbers.... If that is not what you want, probably a bug. We might fix the axes arg, or just use np norm function if that is the case.

**Comment by janden** [comment]
> @janden thoughts on this snippet?

That's very strange. What's the input here? Could be some cancellation that's not happening the way it should be. Looks like `ainner` is using `np.tensordot` underneath, which may do some optimizations that cause problems. One potential solution is running `np.sqrt(np.max(0, …))` to make sure it's positive.

More generally, these functions (`anorm`, `ainner`, and `acorr`) all came about because Matlab doesn't (didn't) have functions for dealing with arrays of higher dimensions than two. Since Numpy seems to be better in this regard, it may make sense to just use those directly.

File: src/aspire/basis/__init__.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ def expand(self, x):
         for isample in range(0, n_data):
             b = self.evaluate_t(x[..., isample])
             # TODO: need check the initial condition x0 can improve the results or not.
-            v[..., isample], info = cg(operator, b, tol=tol)
+            v[..., isample], info = cg(operator, b, tol=tol, atol='legacy')
```
Comment (janden) [inline src/aspire/basis/__init__.py]: Do we know what `atol='legacy'` does? The docs are not very specific here. Will things start breaking if we specify `atol=0`? That setting seems to make sense to me.

**Review by janden — CHANGES_REQUESTED** [review]

**Comment by garrettwrong** [comment]
I dont think forcing the sqrt is right, the problem isn't numerical, I think we are actually computing wrong thing before we give to sqrt...

I think the bug is actually the 2-tuple of indices given to axes. As I read doc, if we give it (0,1) I think it will take along 0 for first array and 1 for the second... In our case x is same array, but we're not computing X[:..] * X[:,...] then...

Yeah I think a `np.linalg.norm(..,axis=)` will do it.  Just checking we didn't intend to do something I'm missing.

File: src/aspire/basis/__init__.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/basis/__init__.py]: Legacy is the default we were using. Let me know what you think, can change...

https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/isolve/iterative.py#L95

Comment (janden) [reply src/aspire/basis/__init__.py]: Legacy behavior looks a bit problematic if `bnrm2` is very small (but not zero). I'm guessing that's why they're changing it. Let's put `atol=0` and see what happens.

Comment (garrettwrong) [reply src/aspire/basis/__init__.py]: SGTM. Tests pass with 0. I think it is fair to revisit if we have convergence problem etc.

**Comment by janden** [comment]
>  As I read doc, if we give it (0,1) I think it will take along 0 for first array and 1 for the second...

But `ainner` is giving `(axes, axes)` to `np.tensordot`, which should give the same axes for both arrays, right?

> Just checking we didn't intend to do something I'm missing.

No it's just supposed to calculate the square root of the sum of squares. Looks like `np.linalg.norm` is similarly deficient here, though:
```
>>> x = np.random.randn(3, 4, 5)
>>> np.linalg.norm(x, axis=(0, 1, 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in norm
  File "/usr/lib/python3/dist-packages/numpy/linalg/linalg.py", line 2563, in norm
    raise ValueError("Improper number of dimensions to norm.")
ValueError: Improper number of dimensions to norm.
```

It can deal with higher-dimensional arrays if you specify `axis=None`, it seems, but that's it.

**Comment by garrettwrong** [comment]
I think tensordot migth want a 2-tuple (i dont think it assumes same axis).  I'll take a look at norm. if I can get what I think the original code wanted... no worries, this is why I wanted to hit the warnings.

**Comment by garrettwrong** [comment]
interestingly, on the case that causes the most trouble:

x.shape (8, 8, 8, 2)
axes (0, 1, 2)

returns a 2x2 matrix :D. I see we do use this code unfortunately.

To use np norm you can specify multipe axis, we just omit the one you are "along". To specify along all, is the none case you mention.

```
>>> x = np.random.randn(3,4,5)
>>> np.linalg.norm(x)
7.21001230307816
>>> np.linalg.norm(x,axis=0).shape
(4, 5)
>>> np.linalg.norm(x,axis=1).shape
(3, 5)
>>> np.linalg.norm(x,axis=2).shape
(3, 4)
>>> np.linalg.norm(x,axis=(0,1)).shape
(5,)
>>> np.linalg.norm(x,axis=(0,2)).shape
(4,)
>>> np.linalg.norm(x,axis=(1,2)).shape
(3,)
>>> np.linalg.norm(x,axis=-1).shape
(3, 4)
>>> y = np.random.randn(3,4,5,6)
>>> np.linalg.norm(y,axis=(0,2)).shape
(4, 6)
```

I am thinking to delete anorm at this point and try to figure out what the invoking code was expecting. How's that sound?

**Comment by garrettwrong** [comment]
I'll make a seperate issue and PR for it I guess.

**Comment by garrettwrong** [comment]
Broke out anorm to separate issue. Otherwise these should be okay for review.

**Review by janden — APPROVED** [review]

**Comment by garrettwrong** [comment]
Hey @junchaoxia , do you want to review this too (it's small)? Thanks

**Review by junchaoxia — APPROVED** [review]
I took a quick look. I think Joakim's review is enough.

**Pull request merged**